pkg/ipcache/metadata.go

```diff
@@ -157,111 +165,155 @@ func (m *metadata) get(prefix string) prefixInfo {
+			if err != nil {
+				// NOTE: This may fail during a 2nd or later
+				// iteration of the loop. To handle this, break
+				// the loop here and continue executing the set
+				// of changes for the prefixes that were
+				// already processed.
+				//
+				// Old identities corresponding to earlier
+				// prefixes may be released as part of this,
+				// so hopefully this forward progress will
+				// unblock subsequent calls into this function.
 				log.WithError(err).WithFields(logrus.Fields{
-					logfields.IPAddr: prefix,
-					logfields.Labels: lbls,
-				}).Error(
-					"Failed to release assigned identity during label injection, this might be a leak.",
-				)
-			} else {
-				log.WithFields(logrus.Fields{
 					logfields.IPAddr:   prefix,
 					logfields.Identity: id,
-				}).Debug(
-					"Released identity to balance reference counts",
+					logfields.Labels:   lbls, // new labels
+				}).Warning(
+					"Failed to allocate new identity while handling change in labels associated with a prefix.",
 				)
+				remainingCIDRs = modifiedCIDRs[i:]
+				err = fmt.Errorf("failed to allocate new identity during label injection: %w", err)
+				break
+			}
+
+			idsToAdd[newID.ID] = lbls.LabelArray()
+			entriesToReplace[prefix] = Identity{
+				ID:     newID.ID,
+				Source: prefixInfo.Source(),
+			}
+		}
+		if entryExists {
+			// 'prefix' is being removed or modified, so some
+			// previous iteration of this code hit one the
+			// 'injectLabels' case above, thereby allocating a
+			// (new) identity. If we delete or update the identity
+			// for 'prefix' in this iteration of the loop, then we
+			// must balance the allocation from the prior
+			// InjectLabels() call by releasing the previous
+			// reference.
+			if _, ok := idsToAdd[id.ID]; !ok {
+				previouslyAllocatedIdentities[prefix] = id
 			}
 		}
 	}
 	// Don't hold lock while calling UpdateIdentities, as it will otherwise run into a deadlock
 	ipc.metadata.Unlock()
 
 	// Recalculate policy first before upserting into the ipcache.
-	if trigger {
-		ipc.UpdatePolicyMaps(context.TODO(), idsToPropagate, nil)
+	if len(idsToAdd) > 0 {
+		ipc.UpdatePolicyMaps(context.TODO(), idsToAdd, nil)
 	}
 
 	ipc.mutex.Lock()
 	defer ipc.mutex.Unlock()
-	for ip, id := range toUpsert {
+	for ip, id := range entriesToReplace {
 		hIP, key := ipc.getHostIPCache(ip)
 		meta := ipc.getK8sMetadata(ip)
-		if _, err := ipc.upsertLocked(ip, hIP, key, meta, Identity{
-			ID:     id.ID,
-			Source: id.Source,
-		}, false /* !force */); err != nil {
-			return fmt.Errorf("failed to upsert %s into ipcache with identity %d: %w", ip, id.ID, err)
+		if _, err2 := ipc.upsertLocked(
+			ip,
+			hIP,
+			key,
+			meta,
+			id,
+			true,
+		); err2 != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.IPAddr:   ip,
+				logfields.Identity: id,
+			}).Error("Failed to replace ipcache entry with new identity after label removal. Traffic may be disrupted.")
```

Comment:
`err2` here?

Reply:
Looks like the sole [failure](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.16-kernel-4.9/1815/testReport/junit/Suite-k8s-1/16/K8sDatapathConfig_Etcd_Check_connectivity/) in CI is hidden here with `err=nil` in the sysdump because of this typo.

Reply:
Ah too bad. I'll fix it up, hopefully the failure is reliable enough that it triggers next time and prints the error.